language/move-lang/src/parser/syntax.rs

```diff
@@ -257,65 +282,80 @@ fn parse_module_name(tokens: &mut Lexer) -> Result<ModuleName, Error> {
 //      ModuleIdent = <Address> "::" <ModuleName>
 fn parse_module_ident(tokens: &mut Lexer) -> Result<ModuleIdent, Error> {
     let start_loc = tokens.start_loc();
-    let address = parse_address(tokens)?;
-    if tokens.peek() != Tok::ColonColon {
-        let unexp_loc = current_token_loc(tokens);
-        let unexp_msg = format!("Unexpected {}", current_token_error_string(tokens));
+    let address = parse_leading_name_access(tokens)?;
 
-        let end_loc = tokens.previous_end_loc();
-        let addr_loc = make_loc(tokens.file_name(), start_loc, end_loc);
-        let exp_msg = "Expected '::' after an address in a module identifier";
-        return Err(vec![
-            (unexp_loc, unexp_msg),
-            (addr_loc, exp_msg.to_string()),
-        ]);
-    }
-
-    consume_token(tokens, Tok::ColonColon)?;
-    let name = parse_module_name(tokens)?.0;
+    consume_access_colon_colon(tokens, start_loc, "after an address in a module identifier")?;
+    let module = parse_module_name(tokens)?;
     let end_loc = tokens.previous_end_loc();
     let loc = make_loc(tokens.file_name(), start_loc, end_loc);
-    Ok(ModuleIdent {
-        locs: (loc, name.loc),
-        value: (address, name.value),
-    })
+    Ok(sp(loc, ModuleIdent_ { address, module }))
 }
 
 // Parse a module access (a variable, struct type, or function):
 //      ModuleAccess =
 //          <Identifier>
```

Comment:
This first option should be `<LeadingNameAccess>`. Or you could write the whole thing as:
```
NameAccessChain =
    <LeadingNameAccess> ( "::" <Identifier> ( "::" <Identifier> )? )?
```